model/fn-execution/src/main/proto/beam_fn_api.proto

```diff
@@ -182,6 +182,48 @@ message ProcessBundleDescriptor {
   org.apache.beam.model.pipeline.v1.ApiServiceDescriptor state_api_service_descriptor = 7;
 }
 
+// Represents a partition of the bundle into two bundles: a "primary" and
+// a "residual", with the following properties:
+// - The work in primary and residual doesn't overlap, and combined, adds up
+//   to the work in the current bundle if the split hadn't happened.
+// - The current bundle, if it keeps executing, will have done none of the
+//   work under residual roots.
+// - The current bundle, if no further splits happen, will have done exactly
+//   the work under primary_roots.
+// For more rigorous definitions see https://s.apache.org/beam-breaking-fusion
+message BundleSplit {
+  // One of the root applications specifying the scope of work for a bundle.
+  message Application {
+    // (Required) The primitive transform to which to pass the element
+    string ptransform_id = 1;
+
+    // (Required) Name of the transform's input to which to pass the element.
+    string input_id = 2;
+
+    // (Required) The encoded element to pass to the transform.
+    bytes element = 3;
+
+    // Approximate lower bounds on timestamps of elements that this PTransform
+    // will produce into each of its output PCollections, when invoked on this
+    // element. Keyed by the transform's local output name.
+    map<string, int64> output_watermarks = 4;
```

Comment:
Do we really want the user to record the output watermark per output?

Wouldn't it make more sense to ask for the input watermark to be returned?
(since the applications themselves represent inputs to a function)

Reply:
No, this has to be output watermark: for an Application of the form `f(x)` , the input watermark would be "a promise about timestamps of future elements from the same PCollection that x came from", but the Application has no control over that - the Application can only make promises about what happens if you execute this Application.

For example, if this is an Application of an SDF that reads data from Kafka, the "element" is (element:(topic, partition), restriction:offset), and there's only 1 output, and we're making a promise about what timestamps will be output if you read this partition from this offset - it's the watermark of this partition; the watermark of the topic/partition PCollection itself is irrelevant.

Reply:
The issue I see is that if you perform a split, how do you hold the input watermark of these applications so that state/side inputs aren't garbage collected, timers don't fire too early (watermark based timers).

It may be that we need both or we could say that the input watermark is always a lower bound for the output watermarks.

The change looks fine for now but I think this is an important point to discuss.